[5.5.x] avoid crashes when working with operational state

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR restructures the way operational environments are created by doing away with command comparisons in favor of creating the environments where they're actually needed. This alleviates the problems with command-specific logic to decide which environment to create for which command early in the process.
Additionally, this backports the change necessary to be able to complete an expand operation from another cluster node.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/1468.
* Requires https://github.com/gravitational/gravity.e/pull/4295.
* Ports https://github.com/gravitational/gravity/pull/1347 (partially).

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
Upgraded a 2-node cluster using the branch.
Attempted the `gravity upgrade --resume` with an active expand operation:
```
$ ./gravity upgrade --resume
no active update operation found
```

a-palchikov committed:
Sync with upstream regarding operation handling.
Improve error message for not found errors for operations.

a-palchikov committed:
Backport changes to use LocalEnvironmentFactory instead of second-guessing the environment based on command used

a-palchikov commented on lib/expand/join.go:
```diff
@@ -433,7 +433,8 @@ func (p *Peer) tryConnect() (op *operationContext, err error) {
 		if trace.IsCompareFailed(err) {
-			p.sendMessage("Waiting for another operation to finish at %v", addr)
+			p.WithError(err).Info("Failed to create expand operation.")
+			p.sendMessage("Waiting to create expand operation at %v: %v", addr, err)
```
The old message was misleading as there multiple possible reasons for this error including cluster being degraded.

a-palchikov commented on tool/gravity/cli/operation.go:
```diff
@@ -210,27 +231,17 @@ func (r *backendOperations) List(localEnv, updateEnv, joinEnv *localenv.LocalEnv
-			} else {
-				log.WithError(err).Warn("Failed to query install operation.")
-			}
+		if err := r.listInstallOperation(); err != nil {
```
Moved it to a separate function.

a-palchikov commented on tool/gravity/cli/system.go:
```diff
@@ -1120,7 +1120,7 @@ func systemUninstall(env *localenv.LocalEnvironment, confirmed bool) error {
 	for _, path := range append(state.StateLocatorPaths, defaults.ModulesPath, defaults.SysctlPath, defaults.GravityEphemeralDir) {
 		// errors are expected since some of them may not exist
-		if err := os.Remove(path); err == nil {
+		if err := os.RemoveAll(path); err == nil {
```
This was not cleaning anything up.

r0mant commented on tool/gravity/cli/clusterupdate.go:
```diff
@@ -97,7 +95,11 @@ func newClusterUpdater(
+	updateEnv, err := environ.NewUpdateEnv()
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
defer updateEnv.Close()? And in the 2 places below as well.

r0mant commented on tool/gravity/cli/operation.go:
```diff
@@ -116,24 +117,44 @@ func completeOperationPlan(localEnv, updateEnv, joinEnv *localenv.LocalEnvironme
 	default:
 		return trace.BadParameter("operation type %q does not support plan completion", op.Type)
 	}
+	if op.Type != ops.OperationInstall && trace.IsNotFound(err) {
```
Is this special use-case mostly to be able to complete join (or any operation?) from another node?

a-palchikov replied:
Yes.

r0mant commented on tool/gravity/cli/operation.go:
```diff
@@ -210,27 +231,17 @@ func (r *backendOperations) List(localEnv, updateEnv, joinEnv *localenv.LocalEnv
 	}
-	if updateEnv != nil {
-		r.getOperationAndUpdateCache(updateEnv.Backend, log.WithField("context", "update"))
+	if err := r.listUpdateOperation(environ); err != nil && !trace.IsNotFound(err) {
```
Hm, I might be misunderstanding the logic here, but the way it was implemented before IIRC we would pass either updateEnv or joinEnv here - one of those would be nil depending on which command is run - and hence we would only query either update or join operation. Now, we query both so backendOperations may contain both upgrade and join operation if both backends are present on the node. Would that confuse anything?

a-palchikov replied:
No, prior to that both `updateEnv` and `joinEnv` could be active at the same time as the `plan` command is ambiguous so it worked similar to this before. 
One thing that to account for is the residual state that might have been left behind as the `leave` (or more specifically `system uninstall`) command would not remove the join state directory (only relevant in 5.5.x though) and since the expand operation was never updated to completion in this local store, it will popup as an unfinished (hence active) and maybe even unrelated to the current cluster.
All because of such disparate state management.

r0mant commented on tool/gravity/cli/operation.go:
```diff
@@ -249,7 +260,49 @@ func (r *backendOperations) init(clusterBackend storage.Backend) error {
+	return nil
+}
+
+func (r *backendOperations) listInstallOperation() error {
```
Nit: This method doesn't return anything so we can do away with the return parameter.

r0mant commented on tool/gravity/cli/operation.go:
```diff
@@ -265,6 +318,16 @@ func (r *backendOperations) getOperationAndUpdateCache(backend storage.Backend,
 }
 
 func (r backendOperations) isActiveInstallOperation() bool {
+	// Bail out if there's an operation from a local backend and we failed to query
```
TBH I'm not sure I follow this comment - could it be reworded somehow maybe? To explain what "r.clusterOperation" is, what "r.operations" are and why this condition means this is not an active install operation.

a-palchikov replied:
I'll update the struct responsible for operation collection.

r0mant commented on tool/gravity/cli/plan.go:
```diff
@@ -64,9 +64,15 @@ func initUpdateOperationPlan(localEnv, updateEnv *localenv.LocalEnvironment) err
 	if err != nil {
+		if trace.IsNotFound(err) {
+			return trace.NotFound(`no operation found.
+This usually means that the installation has failed to start or was not started.
```
Nit: Can this also mean that "join" operation has failed to start or was not started?

a-palchikov replied:
Possible - I'll update the message.

r0mant commented on tool/gravity/cli/plan.go:
```diff
@@ -82,18 +88,23 @@ func displayOperationPlan(localEnv, updateEnv, joinEnv *localenv.LocalEnvironmen
 	default:
 		return trace.BadParameter("unknown operation type %q", op.Type)
 	}
+	if err != nil && trace.IsNotFound(err) {
```
What is this fallback for? Seems like the switch above covers all operation types?

a-palchikov replied:
Looks like a backport gone slightly wrong - I'll fix that.

r0mant commented on tool/gravity/cli/plan.go:
```diff
@@ -109,7 +120,11 @@ func displayClusterOperationPlan(env *localenv.LocalEnvironment, opKey ops.SiteO
+	updateEnv, err := environ.NewUpdateEnv()
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
defer updateEnv.Close()?

r0mant commented on tool/gravity/cli/plan.go:
```diff
@@ -231,4 +228,21 @@ func tryReconcilePlan(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
+		return nil, trace.Wrap(err)
+	}
+	if wizardEnv.Operator == nil {
+		return nil, trace.NotFound("no operation plan")
```
Nit: I think the message here is a bit misleading as we didn't actually check for plan. I'd rather say why the operator is nil here.

a-palchikov replied:
That's a good question - I have no idea how the operator could be nil after a successful call to `NewRemoveEnvironment`.

r0mant commented on tool/gravity/cli/run.go:
```diff
@@ -362,36 +353,58 @@ func Execute(g *Application, cmd string, extraArgs []string) error {
 		}
 		if *g.UpgradeCmd.Phase != "" {
-			return executePhase(localEnv, updateEnv, joinEnv,
+			op, err := getActiveOperation(localEnv, g, "")
```
Nit: Can this logic be moved under "executeUpdatePhase"? Seems a bit weird to have this in run.go.

a-palchikov committed:
Address review comments

knisbet commented on tool/gravity/cli/plan.go:
```diff
@@ -64,9 +64,15 @@ func initUpdateOperationPlan(localEnv, updateEnv *localenv.LocalEnvironment) err
+		if trace.IsNotFound(err) {
+			return trace.NotFound(`no operation found.
+This usually means that the installation/join operation has failed to start or was not started.
+Clean up the node with 'gravity leave' if necessary and start the installation/join with either 'gravity install' or 'gravity join'.
```
How is a user supposed to know if running gravity leave is necessary?

a-palchikov replied:
I'll have the code check and output the appropriate message.

a-palchikov committed:
Address review comments. Add more user-friendly error messages

a-palchikov committed:
Rewrite operation detection to avoid invalid operations previously considered active.

a-palchikov committed:
When displaying operation plan, use the active operation is available,
otherwise use the latest and show multiple operations warning.
Also, remove the use-case of multiple active operations - always use the
latest one if available.
Consider an operation active only if it has not been completed and has a
valid plan.

a-palchikov committed:
Rename opslist -> operationList.
Revert the logic of checking whether the operation is available prior to
updating the cache - this will break for cases when cluster store is
unavailable.

a-palchikov committed:
Simplify last/active operation handling even further:
  * get rid of multiple operations - always use the last if available
  * only consider operations with a plan

a-palchikov committed:
Bump e
